### PR TITLE
Add UnionFields (#3955)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #3955

# Rationale for this change
 

* Reduces cloning of Field #3955
* Adds an abstraction to allow improving ergonomics of interacting with Union datatypes
* Reduces the size of DataType from 56 bytes to 24 bytes

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-schema/src/fields.rs [hunk 1]
```diff
@@ -148,3 +150,91 @@ impl<'de> serde::Deserialize<'de> for Fields {
...
+#[cfg_attr(feature = "serde", serde(transparent))]
+pub struct UnionFields(Arc<[(i8, FieldRef)]>);
+
+impl std::fmt::Debug for UnionFields {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.as_ref().fmt(f)
+    }
+}
+
+impl UnionFields {
```
Comment (tustvold) [inline arrow-schema/src/fields.rs:166]: I purposefully kept the size of this interface down as I don't feel we have very mature support for UnionArrays and therefore I didn't want to paint ourselves into a corner too much w.r.t what this abstraction looked like

File: arrow-schema/src/fields.rs [hunk 2]
```diff
@@ -148,3 +150,91 @@ impl<'de> serde::Deserialize<'de> for Fields {
...
+    }
+
+    /// Returns an iterator over the fields and type ids in this [`UnionFields`]
+    ///
+    /// Note: the iteration order is not guaranteed
+    pub fn iter(&self) -> impl Iterator<Item = (i8, &FieldRef)> + '_ {
+        self.0.iter().map(|(id, f)| (*id, f))
+    }
+
+    pub(crate) fn try_merge(&mut self, other: &Self) -> Result<(), ArrowError> {
```
Comment (alamb) [inline arrow-schema/src/fields.rs:207]: Can you please add some docs about what this does?

**Review by alamb — APPROVED** [review]
> Reduces the size of DataType from 56 bytes to 24 bytes


🎉 

I think given that the size of DataType is an important metric, we should also add a test that verifies it doesn't change by accident -- like https://github.com/apache/arrow-datafusion/blob/8e125d2ecf242b4f4b81f06839900dbb2037cc2a/datafusion/common/src/scalar.rs#L3728-L3738

File: arrow-schema/src/fields.rs [hunk 3]
```diff
@@ -148,3 +150,91 @@ impl<'de> serde::Deserialize<'de> for Fields {
...
+}
+
+impl UnionFields {
+    /// Create a new [`UnionFields`] with no fields
+    pub fn empty() -> Self {
+        Self(Arc::from([]))
+    }
+
+    /// Create a new [`UnionFields`] from a [`Fields`] and array of type_ids
+    pub fn new<F, T>(type_ids: T, fields: F) -> Self
```
[the remark this answers is not in the document]
Comment (tustvold) [reply arrow-schema/src/fields.rs:173]: I honestly wasn't sure about this, the existing code doesn't handle this consistently. I'll file a ticket

File: arrow-schema/src/fields.rs (hunk 2 above)
Comment (tustvold) [reply arrow-schema/src/fields.rs:226]: Sorry, should have called out in review, just moved some code here. Added docs in https://github.com/apache/arrow-rs/pull/3981/commits/07af4ffe9f599e58c0f937bb3e9b6a2844dbec33

**Pull request merged**